### PR TITLE
ListProductViews query

### PR DESCRIPTION
**Pull request by trickreich**

### CODE REVIEW

File: Repository/Product/ProductRepository.php [hunk 1]
```diff
@@ -45,6 +50,69 @@ public function findById(string $id): ?ProductInterface
...
+            $queryBuilder
+                ->join('product.mainCategory', 'mainCategory')
+                ->join(Category::class, 'category')
+                ->andWhere('mainCategory.key IN(:categoryKeys) OR category.key IN(:categoryKeys)')
+                ->setParameter('categoryKeys', $categoryKeys);
+        }
+
+        if ($query) {
+            $queryBuilder
+                ->andWhere('LOWER(product.code) LIKE :query OR LOWER(productInformation.name) LIKE :query')
```
Comment (trickreich) [inline Repository/Product/ProductRepository.php]: ` product.code` and ` productInformation.name`  should come from config

File: Model/Product/Handler/Message/PublishProductMessageHandler.php [hunk 2]
```diff
@@ -128,5 +137,34 @@ private function publishProductInformation(ProductInterface $product, string $lo
...
+    private function publishProductInformationAttributeValues(
+        ProductInformationInterface $draftProductInformation,
+        ProductInformationInterface $liveProductInformation
+    ) {
+        // process existing and new
+        $processedAttributeValueCodes = [];
+        foreach ($draftProductInformation->getAttributeValues() as $draftAttributeValue) {
+            $attributeValue = $liveProductInformation->findAttributeValueByCode($draftAttributeValue->getCode());
+            if (!$attributeValue) {
+                // create new one
```
Comment (wachterjohannes) [inline Model/Product/Handler/Message/PublishProductMessageHandler.php]: remove

File: Model/Product/Handler/Query/ListProductViewsQueryHandler.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+use Sulu\Bundle\SyliusConsumerBundle\Model\Dimension\DimensionRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\ProductInformationAttributeValueRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\ProductRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\Query\ListProductViewsQuery;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\View\ProductViewFactoryInterface;
+
+class ListProductViewsQueryHandler
+{
+    const DEFAULT_PAGE_SIZE = 10;
+    const DEFAULT_QUERY_FIELDS = ['product.code', 'productInformation.name'];
```
Comment (wachterjohannes) [inline Model/Product/Handler/Query/ListProductViewsQueryHandler.php]: why not default in query-message?

File: Model/Product/Handler/Query/ListProductViewsQueryHandler.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+use Sulu\Bundle\SyliusConsumerBundle\Model\Dimension\DimensionInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Dimension\DimensionRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\ProductInformationAttributeValueRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\ProductRepositoryInterface;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\Query\ListProductViewsQuery;
+use Sulu\Bundle\SyliusConsumerBundle\Model\Product\View\ProductViewFactoryInterface;
+
+class ListProductViewsQueryHandler
+{
+    const DEFAULT_PAGE_SIZE = 10;
```
Comment (wachterjohannes) [inline Model/Product/Handler/Query/ListProductViewsQueryHandler.php]: why not default in query-message?

File: Model/Product/ProductInformationAttributeValue.php [hunk 5]
```diff
@@ -82,6 +82,11 @@ public function __construct(ProductInformationInterface $productInformation, str
         $this->type = $type;
     }
 
+    public static function getGetterByType(string $type): string
```
Comment (wachterjohannes) [inline Model/Product/ProductInformationAttributeValue.php]: `getPropertyNameByType`

File: Model/Product/Handler/Query/ListProductViewsQueryHandler.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                DimensionInterface::ATTRIBUTE_KEY_LOCALE => $query->getLocale(),
+            ]
+        );
+
+        $attributeFilters = [];
+        foreach ($query->getAttributeFilters() as $attributeCode => $attributeValue) {
+            $attributeFilters[] = [
+                'code' => $attributeCode,
+                'value' => $attributeValue,
+                'type' => $this->productInformationAttributeValueRepository->getTypeByCode($attributeCode),
```
Comment (wachterjohannes) [inline Model/Product/Handler/Query/ListProductViewsQueryHandler.php]: i would preload all types: 

```
$types = $this->productInformationAttributeValueRepository->getTypeByCodes(array_keys($query->getAttributeFilters()));
```

should return this:

```
[
    'attribute_code_1' => 'type1',
    ...
]
```

File: Repository/Product/ProductRepository.php [hunk 7]
```diff
@@ -45,6 +48,56 @@ public function findById(string $id): ?ProductInterface
...
+        if ($categoryKeys) {
+            $queryBuilder
+                ->join('product.mainCategory', 'mainCategory')
+                ->join(Category::class, 'category')
+                ->andWhere('mainCategory.key IN(:categoryKeys) OR category.key IN(:categoryKeys)')
+                ->setParameter('categoryKeys', $categoryKeys);
+        }
+
+        foreach ($attributesFilter as $attributeId => $attributeValue) {
+            $placeholderId = $attributeId . '_' . 'id';
```
[the remark this answers is not in the document]
Comment (wachterjohannes) [reply Repository/Product/ProductRepository.php]: :+1:

**Pull request merged**